In [2]:
import os
import json
import numpy as np

### Find total num of subjects and file path for creating config.json later 

In [14]:
# Get current working directory
current_dir = os.getcwd()

# Go into "data" folder and list all entries
app_dir = current_dir+"/data"
entries = os.listdir(app_dir)

# Initialize an empty list to store folder path and subjectID
folders = []
subjectID = []

# Iterate over all entries
LENGTH = len("sub-")
for entry in entries:
    # Check if the entry is a directory
    if os.path.isdir(os.path.join(app_dir, entry)):
        # If it is, add it to the list of folders
        subjectID.append(int(entry[LENGTH:]))
        folders.append("../data/" + entry + "/cortexmap")

# sort folder names alphabetically
folders.sort()

# Print num of folder to verify
print("Total number of folders: ", len(folders))

Total number of folders:  1062


### Generate boostrap samples

- Number of subjects = 1062
- Number of bootstrap samples = 1000
- So, there are 1000 samples. Each sample has 1062 input files (sampled with replacement)
- Initial seed: 3212024171403

In [24]:
initial_seed = 3212024171403 # initial seed: 3212024171403

# Generate random seeds
rng = np.random.default_rng(initial_seed)
num_samples = 1000 # num of bootstrap samples = num of seeds
seeds = rng.choice(1000000000000000000, num_samples, replace=False) # includes 1 seed per sample

# Generate bootstrap samples
samples = [] # 2D arr. 1 row = 1 boostrap sample. 1 element = 1 subject = index folders arr

for seed in seeds:
    rng = np.random.default_rng(seed)
    total_subjects = len(folders) # total number of subjects = size of sample
    samples.append(rng.choice(total_subjects, total_subjects))

# print dimension of samples to verify
print("dimensions of bootstrap sample:")
print(len(samples))
print(len(samples[0]))

dimensions of bootstrap sample:
1000
1062


### Create config.json file for fsaverage app

In [25]:
# Create list of folder paths for the 1000 bootstrap samples
counter = 1
for sample in samples:
    subjects = [] # folder paths for 1 boostrap sample
    for i in sample:
        subjects.append(folders[i])

    print("Number of input files: ", len(subjects)) # printing to confirm
    # Create a dictionary with "input" key (threshold and dilate set to 0)
    json_data = {"threshold" : 0,
                "dilate" : 0,
                "inputs": subjects}

    # Specify the output JSON file path (outputs to current directory)
    output_file_path = "config_s" + str(counter) + ".json"

    # Write data to JSON file
    with open(output_file_path, 'w') as f:
        json.dump(json_data, f, indent=4)

    print(f"Data has been written to {output_file_path}")
    counter += 1

Number of input files:  1062
Data has been written to config_s1.json
Number of input files:  1062
Data has been written to config_s2.json
Number of input files:  1062
Data has been written to config_s3.json
Number of input files:  1062
Data has been written to config_s4.json
Number of input files:  1062
Data has been written to config_s5.json
Number of input files:  1062
Data has been written to config_s6.json
Number of input files:  1062
Data has been written to config_s7.json
Number of input files:  1062
Data has been written to config_s8.json
Number of input files:  1062
Data has been written to config_s9.json
Number of input files:  1062
Data has been written to config_s10.json
Number of input files:  1062
Data has been written to config_s11.json
Number of input files:  1062
Data has been written to config_s12.json
Number of input files:  1062
Data has been written to config_s13.json
Number of input files:  1062
Data has been written to config_s14.json
Number of input files:  1062
